include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+
+        inline void
+        publish_dataset(DataSet<T> &&data) {
+            if constexpr (!std::is_same_v<Callback, null_type>) {
+                callback(std::move(data));
+            } else {
+                auto poller = polling_handler.lock();
+                if (!poller) {
+                    this->set_expired();
+                    return;
+                }
+
+                auto write_data = poller->writer.reserve_output_range(1);
+                if (block) {
+                    write_data[0] = std::move(data);
+                    write_data.publish(1);
+                } else {
+                    if (poller->writer.available() > 0) {
+                        write_data[0] = std::move(data);
+                        write_data.publish(1);
+                    } else {
+                        poller->drop_count++;
+                    }
+                }
+            }
+        }
+
+        void
+        set_sample_rate(float r) override {
+            sample_delay = std::round(std::chrono::duration_cast<std::chrono::duration<float>>(time_delay).count() * r);
+            // TODO do we need to update the requested_samples of pending here? (considering both old and new time_delay)
+        }
+
+        void
+        process(std::span<const T>, std::span<const T> in_data, std::optional<property_map> tag_data0) override {
+            if (tag_data0 && trigger_predicate({ 0, *tag_data0 })) {
+                auto new_pending = pending_snapshot{ *tag_data0, sample_delay, sample_delay };
+                // make sure pending is sorted by number of pending_samples (insertion might be not at end if sample rate decreased; TODO unless we adapt them in set_sample_rate, see there)
+                auto rit = std::find_if(pending.rbegin(), pending.rend(), [delay = sample_delay](const auto &other) { return other.pending_samples < delay; });
+                pending.insert(rit.base(), std::move(new_pending));
+            }
+
+            auto it = pending.begin();
+            while (it != pending.end()) {
+                if (it->pending_samples >= in_data.size()) {
+                    it->pending_samples -= in_data.size();
+                    break;
+                }
+
+                DataSet<T> dataset;
+                dataset.timing_events = { { { -static_cast<tag_t::signed_index_type>(it->delay), std::move(it->tag_data) } } };
+                dataset.signal_values = { in_data[it->pending_samples] };
+                publish_dataset(std::move(dataset));
+
+                it = pending.erase(it);
+            }
+        }
+
+        void
+        flush() override {
+            pending.clear();
+            if (auto p = polling_handler.lock()) {
+                p->finished = true;
+            }
+        }
+    };
+};
+
+} // namespace fair::graph
+
+ENABLE_REFLECTION_FOR_TEMPLATE_FULL((typename T), (fair::graph::data_sink<T>), in, sample_rate);
```

Comment:
Missing fields that need to be registered:
  * signal_name
  * signal_unit
  * signal_min
  * signal_max
  
  These would also need to be propagated to the poller/callback function and populating the DataSet<T> meta information

Reply:
Add to reflection, not populated/propagated yet

Reply:
Now read from settings or incoming tags (settings have preference), and populating the datasets.